### PR TITLE
WIP [17.07] update vendor of components/engine in cli

### PR DESCRIPTION
**Pull request by andrewhsu**
update the `components/cli/vendor/github.com/docker/docker` to vendor from `components/engine` by running `make vendor`

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

### CODE REVIEW

File: components/engine/client/client.go [hunk 1]
```diff
@@ -261,13 +261,8 @@ func (cli *Client) NegotiateAPIVersionPing(p types.Ping) {
...
 	}
 
-	// if the client is not initialized with a version, start with the latest supported version
-	if cli.version == "" {
-		cli.version = api.DefaultVersion
-	}
-
-	// if server version is lower than the maximum version supported by the Client, downgrade
-	if versions.LessThan(p.APIVersion, api.DefaultVersion) {
+	// if server version is lower than the current cli, downgrade
```
Comment (vieux) [inline components/engine/client/client.go]: @tonistiigi are we sure about this one ?

Comment (tiborvass) [inline components/engine/client/client.go]: @vieux yes

File: components/cli/vendor/github.com/docker/docker/api/types/events/events.go [hunk 2]
```diff
@@ -19,6 +19,8 @@ const (
 	NodeEventType = "node"
 	// SecretEventType is the event type that secrets generate
 	SecretEventType = "secret"
+	// ConfigEventType is the event type that configs generate
+	ConfigEventType = "config"
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/api/types/events/events.go]: Brought in by:
* https://github.com/moby/moby/pull/34032 make engine support cluster config event

File: components/cli/vendor/github.com/docker/docker/api/types/swarm/runtime/gen.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+//go:generate protoc -I . --gogofast_out=import_path=github.com/docker/docker/api/types/swarm/runtime:. plugin.proto
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/api/types/swarm/runtime/gen.go]: Brought in by this PR:
* https://github.com/moby/moby/pull/33575 Plugins on swarm

as well as the following few diffs

File: components/cli/vendor/github.com/docker/docker/api/types/swarm/secret.go [hunk 4]
```diff
@@ -12,7 +12,8 @@ type Secret struct {
 // SecretSpec represents a secret specification from a secret in swarm
 type SecretSpec struct {
 	Annotations
-	Data []byte `json:",omitempty"`
+	Data   []byte  `json:",omitempty"`
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/api/types/swarm/secret.go]: Brought in by this PR:
* https://github.com/moby/moby/pull/34123 pluggable secret backend

File: components/cli/vendor/github.com/docker/docker/api/types/swarm/task.go [hunk 5]
```diff
@@ -1,6 +1,10 @@
 package swarm
 
-import "time"
+import (
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/api/types/swarm/task.go]: Also brought in by PR:
* https://github.com/moby/moby/pull/33575 Plugins on swarm

File: components/cli/vendor/github.com/docker/docker/client/errors.go [hunk 6]
```diff
@@ -228,7 +228,7 @@ func IsErrPluginPermissionDenied(err error) bool {
 // NewVersionError returns an error if the APIVersion required
 // if less than the current supported version
 func (cli *Client) NewVersionError(APIrequired, feature string) error {
-	if versions.LessThan(cli.version, APIrequired) {
+	if cli.version != "" && versions.LessThan(cli.version, APIrequired) {
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/client/errors.go]: Brought in by PR:
* https://github.com/moby/moby/pull/33887 Fix NewVersionError() for clients using default version

File: components/cli/vendor/github.com/docker/docker/client/service_create.go [hunk 7]
```diff
@@ -6,9 +6,9 @@ import (
 
 	"github.com/docker/distribution/reference"
 	"github.com/docker/docker/api/types"
-	registrytypes "github.com/docker/docker/api/types/registry"
 	"github.com/docker/docker/api/types/swarm"
 	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/client/service_create.go]: This change was brought in with PR:
* https://github.com/moby/moby/pull/33575 Plugins on swarm

File: components/engine/client/client.go (hunk 1 above)
Comment (tonistiigi) [inline components/engine/client/client.go]: this is tracked by https://github.com/moby/moby/pull/34511

File: components/cli/vendor/github.com/docker/docker/client/session/filesync/diffcopy.go [hunk 8]
```diff
@@ -9,9 +9,10 @@ import (
 	"github.com/tonistiigi/fsutil"
 )
 
-func sendDiffCopy(stream grpc.Stream, dir string, excludes []string, progress progressCb) error {
+func sendDiffCopy(stream grpc.Stream, dir string, includes, excludes []string, progress progressCb) error {
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/client/session/filesync/diffcopy.go]: Brought in with this PR:
* https://github.com/moby/moby/pull/33859 Add path filtering to build session client

File: components/cli/vendor/github.com/docker/docker/pkg/archive/archive.go [hunk 9]
```diff
@@ -93,6 +93,16 @@ const (
 	OverlayWhiteoutFormat
 )
 
+const (
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/archive/archive.go]: Brought in with PR:
* https://github.com/moby/moby/pull/33935 pkg/archive.FileInfoHeader: fill file type bits

File: components/cli/vendor/github.com/docker/docker/pkg/archive/archive_linux.go [hunk 10]
```diff
@@ -5,9 +5,9 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
-	"syscall"
 
 	"github.com/docker/docker/pkg/system"
+	"golang.org/x/sys/unix"
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/archive/archive_linux.go]: Brought in with this PR:
* https://github.com/moby/moby/pull/33399 [project] Switch most syscalls to golang.org/x/sys

File: components/cli/vendor/github.com/docker/docker/pkg/homedir/homedir_unix.go [hunk 11]
```diff
@@ -1,18 +1,16 @@
+// +build !windows
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/homedir/homedir_unix.go]: Brought in with this PR:
* https://github.com/moby/moby/pull/34040 Split homedir files by operating system

File: components/cli/vendor/github.com/docker/docker/pkg/system/xattrs_linux.go [hunk 12]
```diff
@@ -1,63 +1,29 @@
...
-		dataBytes = unsafe.Pointer(&data[0])
-	} else {
-		dataBytes = unsafe.Pointer(&_zero)
-	}
-	_, _, errno := syscall.Syscall6(syscall.SYS_LSETXATTR, uintptr(unsafe.Pointer(pathBytes)), uintptr(unsafe.Pointer(attrBytes)), uintptr(dataBytes), uintptr(len(data)), uintptr(flags), 0)
-	if errno != 0 {
-		return errno
-	}
-	return nil
+	return unix.Lsetxattr(path, attr, data, flags)
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/system/xattrs_linux.go]: Brought in by PR:
* https://github.com/moby/moby/pull/33539 Use symlink xattr functions from x/sys/unix

File: components/cli/vendor/github.com/docker/docker/pkg/term/termios_linux.go [hunk 13]
```diff
@@ -18,20 +16,21 @@ type Termios unix.Termios
 // mode and returns the previous state of the terminal so that it can be
 // restored.
 func MakeRaw(fd uintptr) (*State, error) {
-	var oldState State
-	if _, _, err := unix.Syscall(unix.SYS_IOCTL, fd, getTermios, uintptr(unsafe.Pointer(&oldState.termios))); err != 0 {
+	termios, err := unix.IoctlGetTermios(int(fd), getTermios)
```
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/term/termios_linux.go]: Brought in by PR:
* https://github.com/moby/moby/pull/34135 Replace manually written wrappers by functions from x/sys/unix

**Comment by tiborvass** [comment]
we cool yo

LGTM

**Comment by vieux** [comment]
LGTM

**Comment by andrewhsu** [comment]
Had to add commit to re-vendor golang.org/x/sys to fix errors on compile. Progressed further but still seeing errors:
```
[init] # github.com/docker/cli/cli/command/service
[init] cli/command/service/opts.go:617: cannot use swarm.ContainerSpec literal (type swarm.ContainerSpec) as type *swarm.ContainerSpec in field value
[init] cli/command/service/update.go:284: cspec.Labels undefined (type **swarm.ContainerSpec has no field or method Labels)
[init] cli/command/service/update.go:285: cspec.Image undefined (type **swarm.ContainerSpec has no field or method Image)
[init] cli/command/service/update.go:286: cspec.Args undefined (type **swarm.ContainerSpec has no field or method Args)
[init] cli/command/service/update.go:287: cspec.Command undefined (type **swarm.ContainerSpec has no field or method Command)
[init] cli/command/service/update.go:288: cspec.Env undefined (type **swarm.ContainerSpec has no field or method Env)
[init] cli/command/service/update.go:289: cspec.Dir undefined (type **swarm.ContainerSpec has no field or method Dir)
[init] cli/command/service/update.go:290: cspec.User undefined (type **swarm.ContainerSpec has no field or method User)
[init] cli/command/service/update.go:291: cspec.Hostname undefined (type **swarm.ContainerSpec has no field or method Hostname)
[init] cli/command/service/update.go:292: cspec.Mounts undefined (type **swarm.ContainerSpec has no field or method Mounts)
[init] cli/command/service/update.go:292: too many errors
[init] make: *** [Makefile:28: binary] Error 2
```

Will revendor swarmkit to line up with what components/engine uses.

**Comment by dnephin** [comment]
This issue can probably be closed?

As we discussed a little while ago it doesn't make sense to try and sync vendor. We should use the vendor that comes with the component and only update them by cherry-picking fixes from the upstream component repo.

**Comment by vieux** [comment]
I agree we should close this.

**Pull request closed without merging**